crates/task-impls/src/quorum_proposal.rs

```diff
@@ -0,0 +1,501 @@
+use async_lock::RwLock;
+use hotshot_task::{
+    dependency::{AndDependency, EventDependency, OrDependency},
+    dependency_task::{DependencyTask, HandleDepOutput},
+};
+use hotshot_types::{
+    consensus::Consensus,
+    data::{QuorumProposal, ViewChangeEvidence},
+    event::Event,
+    message::Proposal,
+    simple_certificate::QuorumCertificate,
+    traits::{
+        block_contents::BlockHeader,
+        network::{ConnectedNetwork, ConsensusIntentEvent},
+        node_implementation::{ConsensusTime, NodeImplementation, NodeType},
+    },
+    vote::HasViewNumber,
+};
+
+#[cfg(async_executor_impl = "async-std")]
+use async_std::task::JoinHandle;
+#[cfg(async_executor_impl = "tokio")]
+use tokio::task::JoinHandle;
+use tracing::{debug, error, instrument};
+
+use crate::{
+    consensus::CommitmentAndMetadata,
+    events::HotShotEvent,
+    helpers::{broadcast_event, cancel_task},
+};
+
+/// Validate a quorum proposal.
+#[allow(clippy::needless_pass_by_value)]
+fn validate_quorum_proposal<TYPES: NodeType>(
+    _quorum_proposal: Proposal<TYPES, QuorumProposal<TYPES>>,
+    _event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+) -> bool {
+    true
+}
+
+/// Validate a quorum proposal.
+#[allow(clippy::needless_pass_by_value)]
+fn validate_view_change_evidence<TYPES: NodeType>(
+    _certificate: ViewChangeEvidence<TYPES>,
+    _event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+) -> bool {
+    true
+}
+
+/// Validate a quorum proposal.
+#[allow(clippy::needless_pass_by_value)]
+fn validate_quorum_certificate<TYPES: NodeType>(
+    _certificate: QuorumCertificate<TYPES>,
+    _event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+) -> bool {
+    true
+}
+
+/// Validate a quorum proposal.
+#[allow(clippy::needless_pass_by_value)]
+fn validate_payload_and_metadata<TYPES: NodeType>(
+    _payload_commitment_and_metadata: CommitmentAndMetadata<<TYPES as NodeType>::BlockPayload>,
+    _event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+) -> bool {
+    true
+}
+
+/// Proposal dependency types. These types represent events that precipitate a proposal.
+#[derive(PartialEq)]
+enum ProposalDependency {
+    /// For the `SendPayloadCommitmentAndMetadata` event.
+    PayloadAndMetadata,
+
+    /// For the `ViewSyncFinalizeCertificate2Recv` event.
+    ViewSync,
+
+    /// For the `QuorumProposalRecv` event.
+    QuorumProposal,
+
+    /// For the `QCFormed` event.
+    QCFormed,
+}
+
+/// Handler for the proposal dependency
+struct ProposalDependencyHandle<TYPES: NodeType> {
+    /// The view number to propose for.
+    view_number: TYPES::Time,
+
+    /// The event sender.
+    sender: Sender<Arc<HotShotEvent<TYPES>>>,
+
+    /// Reference to consensus. The replica will require a write lock on this.
+    #[allow(dead_code)]
+    consensus: Arc<RwLock<Consensus<TYPES>>>,
+}
+impl<TYPES: NodeType> HandleDepOutput for ProposalDependencyHandle<TYPES> {
+    type Output = Vec<Arc<HotShotEvent<TYPES>>>;
+
+    #[allow(clippy::no_effect_underscore_binding)]
```

Comment:
This'll be gone when I add validation here and actually use these. In particular, I would like feedback on this structure. I _hate_ mutating, but this seemed like the best way to unpack multiple heterogeneous types without engineering myself into a black hole. I think I'll split this out into a validation function since not all of these will be set in a given proposal event.

Reply:
I don't like mutating in general either. This is probably fine for this function though, since we only mutate each value once.